### PR TITLE
Kartytko assignment2

### PR DESCRIPTION
**Pull request by kartytko**
Adding solutions for second assignment :)

### CODE REVIEW

File: kartytko/assignment-2/Node.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            answer += "null";
+        }
+        answer += "; right: ";
+        if(rightChild != null){
+            answer += rightChild.value;
+        } else {
+            answer += "null";
+        }
+        answer += "; parent: ";
+        if(parent != null){
```
Comment (ltache) [inline kartytko/assignment-2/Node.java]: Nit: I'd suggest you to be consistent with code formatting (i.e having exactly one space after brackets ") {"). Do you use any IDE in particular? I'm asking because its auto format might help here.

File: kartytko/assignment-2/Tree.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+import java.util.Random;
+import java.util.Stack;
+
+public class Tree {
+    private int treeSize = 0;
+    private Node root;
+    public Tree () {
+        root = null;
+    }
+    public Tree (Node n) { root = n; treeSize++; }
```
Comment (ltache) [inline kartytko/assignment-2/Tree.java]: Nit: It's recommended to have only one instruction per line because the code is easier to read that way.

File: kartytko/assignment-2/Tree.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+    // Create node with given value and insert it into random place in tree.
+    public boolean addNodeRandomly(int newNodeValue) {
```
Comment (ltache) [inline kartytko/assignment-2/Tree.java]: I really like that you created this random tree generator. Very nice!

File: kartytko/assignment-2/Tree.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+                    nodeIterator.setLeftChild(newNode);
+                    break;
+                }
+                nodeIterator = nodeIterator.getLeftChild();
+            }
+        }
+        return true;
+    }
+
+    public Node findNode(int soughtValue) {
```
Comment (ltache) [inline kartytko/assignment-2/Tree.java]: Very nice! I like that you implemented the iterative solution :)

File: kartytko/assignment-2/Tree.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            return null;
+        }
+
+        Node node1Iterator = node1;
+        Node node2Iterator = node2;
+
+        // Not storing values results with n^2 complexity.
+        while(node1Iterator.getParent() != null) {
+            while(node2Iterator.getParent() != null) {
+                if(node2Iterator.getParent().getValue() == node1Iterator.getParent().getValue()){
```
Comment (ltache) [inline kartytko/assignment-2/Tree.java]: I think here could be a problem if the nodes given as input are identical. That means node1 = node2 and then you start looking at the parents which will be equal obviously and you'll return the parent in that case (but the lowest common ancestor is the node itself).
What do you think?

File: kartytko/assignment-2/Tree.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+            }
+            node1Iterator = node1Iterator.getParent();
+            node2Iterator = node2;
+        }
+
+        return root;
+    }
+
+
+    public List<Node> findTheAncestors (int soughtValue) {
```
Comment (ltache) [inline kartytko/assignment-2/Tree.java]: Very nice. Could you think of also at a solution that doesn't use the link to the parent?
You don't necessarily have to implement it in this assignment but I think it might be useful as they don't give you the link to the parent that often in an interview.

File: kartytko/assignment-2/TreeTest.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        node3.setLeftChild(node4);
+        node4.setParent(node3);
+        Tree tree = new Tree(node1);
+
+        Node foundNode = tree.findNode(10);
+        Assert.assertEquals(null, foundNode);
+    }
+
+    @Test
+    public void insertNodeWithNotUniqueKey(){
```
Comment (ltache) [inline kartytko/assignment-2/TreeTest.java]: Nice! Up to this point I was wondering where's the size used.

File: kartytko/assignment-2/Tree.java [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+                } else {
+                    stack.push(rightChild);
+                }
+            }
+        }
+        // Return null if value not found in entire tree.
+        return null;
+    }
+
+    public void generateRandomNodes(int nodesAmount) {
```
Comment (ltache) [inline kartytko/assignment-2/Tree.java]: Nit: From what I see, this method is only used in the unit-tests. Would it be possible to move it there?

File: kartytko/assignment-2/TreeTest.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+    @Test
+    public void findAncestorsOfNodeAbsentInTree(){
+        Tree tree = new Tree();
+        tree.generateRandomNodes(10);
+        Assert.assertEquals(null, tree.findTheAncestors(11));
+    }
+
+    @Test
+    public void findCommonAncestorsExampleFromAssignment(){
```
Comment (ltache) [inline kartytko/assignment-2/TreeTest.java]: I'd also add some corner-cases here. For example the lowest common ancestor if the given nodes are the same node.

File: kartytko/assignment-2/Tree.java (hunk 2 above)
Comment (kartytko) [reply kartytko/assignment-2/Tree.java]: Sure! My bad :)

File: kartytko/assignment-2/Tree.java (hunk 5 above)
Comment (kartytko) [reply kartytko/assignment-2/Tree.java]: I added sightly different logic in my new commit. Now I compare node2Iterator.getValue() == node1Iterator.getValue(), so it covers example given by you.

File: kartytko/assignment-2/TreeTest.java (hunk 9 above)
Comment (kartytko) [reply kartytko/assignment-2/TreeTest.java]: Sure! I added tree more:
findCommonAncestorForTheSameNode
findCommonAncestorsForAbsentNodes
findCommonAncestorsForRoot

File: kartytko/assignment-2/Tree.java (hunk 8 above)
Comment (kartytko) [reply kartytko/assignment-2/Tree.java]: You're right! It makes more sense to have it in TreeTest.java. I moved it there.

**Comment by kartytko** [comment]
Thank you, Lavinia!
I updated the code and tried to cover all requested changes. Let me know, if there is anything else I could change/improve :)
